File: ruby/Move.rb
#  Move.rb
#  
#  Copyright 2017 Bert <luke@luke-desktop>
#  
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#  
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA 02110-1301, USA.
#  
#  

require "KingPiece"

class Move
	attr_reader :moveFrom, :moveTo
	
	def initialize(piece, board, moveFrom, moveTo)
		@piece, @board, @moveFrom, @moveTo = piece, board, moveFrom, moveTo
	end
	
	def Move.simple_move(board, piece, startCoord, endCoord)
        return SimpleMove.new(piece, board, startCoord, endCoord)
	end
    
    
	def Move.jump_move(board, piece, startCoord, endCoord)
        return JumpMove.new(piece, board, startCoord, endCoord, Coordinate.extending(startCoord, endCoord))
	end
    
    def Move.multi_jump_move(board, piece, startCoord, endCoord, move)
        return MultiJumpMove.new(piece, board, startCoord, endCoord, Coordinate.extending(startCoord, endCoord), move)
    end
    
	def move_and_king_piece(piece)
		@board[@moveTo] = piece
        if @board.final_row_for_piece?(piece, @moveTo) && piece.can_be_kinged?
            @board[@moveTo] = KingPiece.new(piece.color)
        end
	end
	
	def to_s
		@piece.to_s + " from " + moveFrom.to_s + " to " + moveTo.to_s
	end
end

class SimpleMove < Move
	def initialize(piece, board, moveFrom, moveTo)
		super(piece,board,moveFrom,moveTo)
	end
	
	def execute
		@board.remove(@moveFrom)
		move_and_king_piece(@piece)
	end
	
	def unExecute
		@board.remove(@moveTo)
		@board[@moveFrom] = @piece
	end

end

class JumpMove < Move
	def initialize(piece, board, moveFrom, existingPieceLocation, moveTo)
		super(piece,board,moveFrom,moveTo)
		@removed_piece = nil
		@existing_piece_location = existingPieceLocation
	end
	
	def execute
		if @removed_piece == nil
			@removed_piece = @board[@existing_piece_location]
			@board.remove(@moveFrom)
			move_and_king_piece(@piece)
			@board.remove(@existing_piece_location)
		else
			raise ("Trying to re-execute jump move " + self.to_s)
		end
	end
	
	def unExecute
		if @removed_piece != nil
			@board.remove(@moveTo)
			@board[@moveFrom] = @piece
			@board[@existing_piece_location] = @removed_piece
			@removed_piece = nil
		else
			raise ("TTrying to unexecute jump move that has not been executed " + @self)
		end
	end
	
	def to_s
		return super().to_s() + " jumping " + @existing_piece_location.to_s
	end
end

class MultiJumpMove < JumpMove
	def initialize(piece, board, moveFrom, existingPieceLocation, moveTo, move)
		super(piece,board,moveFrom,existingPieceLocation,moveTo)
		@previous_move = move
	end

	def execute
		@previous_move.execute
		super
	end
	
	def unExecute
		super
		@previous_move.unExecute()
	end
	
	def to_s
		return  @previous_move.to_s + "\n\t" + super().to_s()
	end
end

File: python/python/src/Move.py
'''
Created on Apr 9, 2017

@author: luke
'''

from Coordinate import Coordinate
from KingPiece import KingPiece

class Move(object):
    def __init__(self, piece, board, moveFrom, moveTo):
        self.piece = piece
        self.board = board
        self.moveFrom = moveFrom
        self.moveTo = moveTo
        
    def __repr__(self):
        return "{} moving from {} to {}".format(self.piece, self.moveFrom, self.moveTo)
    
    @staticmethod
    def simpleMove(board, piece, startCoord, endCoord):
        return SimpleMove(piece, board, startCoord, endCoord)
    
    @staticmethod
    def jumpMove(board, piece, startCoord, endCoord):
        return JumpMove(piece, board, startCoord, endCoord, Coordinate.extending(startCoord, endCoord))
    
    @staticmethod
    def multiJumpMove(board, piece, startCoord, endCoord, move):
        return MultiJumpMove(piece, board, startCoord, endCoord, Coordinate.extending(startCoord, endCoord), move)
    
    def moveAndKingPiece(self, piece):
        self.board.placePieceAt(piece, self.moveTo)
        if self.board.isFinalRowForPiece(piece, self.moveTo) and piece.canBeKinged:
            self.board.placePieceAt(KingPiece(piece.color), self.moveTo)

class SimpleMove(Move):
    def __init__(self, piece, board, moveFrom, moveTo):
        super().__init__(piece, board, moveFrom, moveTo)
    
    def execute(self):
        del self.board[self.moveFrom]
        self.moveAndKingPiece(self.piece)
    
    def unExecute(self):
        del self.board[self.moveTo]
        self.board[self.moveFrom] = self.piece
        
    
class JumpMove(Move):
    def __init__(self, piece, board, moveFrom, existingPieceLocation, moveTo):
        super().__init__(piece, board, moveFrom, moveTo)
        self.removedPiece = None
        self.existingPieceLocation = existingPieceLocation
      
    def execute(self):
        if self.removedPiece is None:
            self.removedPiece = self.board[self.existingPieceLocation]
            del self.board[self.moveFrom]
            self.moveAndKingPiece(self.piece)
            del self.board[self.existingPieceLocation]
        else:
            raise RuntimeError("Trying to re-execute jump move " + str(self))
        
    def unExecute(self):
        if self.removedPiece is not None:
            del self.board[self.moveTo]
            self.board[self.moveFrom] = self.piece
            self.board[self.existingPieceLocation] = self.removedPiece
            self.removedPiece = None
        else:
            raise RuntimeError("Trying to unexecute jump move that has not been executed" + str(self))

class MultiJumpMove(JumpMove):
    def __init__(self, piece, board, moveFrom, existingPieceLocation, moveTo, previousMove):
        super().__init__(piece, board, moveFrom, existingPieceLocation, moveTo)
        self.previousMove = previousMove
        
    def execute(self):
        self.previousMove.execute()
        JumpMove.execute(self)
        
    def unExecute(self):
        JumpMove.unExecute(self)
        self.previousMove.unExecute()
